nexus/src/app/background/tasks/instance_reincarnation.rs

```diff
@@ -0,0 +1,715 @@
+                            "instance_id" => %instance_id,
+                            "last_reincarnated_at" => %last_reincarnation,
+                        );
+                        continue;
+                    }
+                }
+
+                let prepared_saga = instance_start::SagaInstanceStart::prepare(
+                    &instance_start::Params {
+                        db_instance,
+                        serialized_authn: serialized_authn.clone(),
+                        reason: instance_start::Reason::AutoRestart,
+                    },
+                );
+                match prepared_saga {
+                    Ok(saga) => {
+                        let start_saga = self.sagas.clone();
+                        tasks.spawn(async move {
+                            start_saga
+                                .saga_start(saga)
+                                .await
+                                .map_err(|e| (instance_id, e))?;
+                            Ok(instance_id)
+                        });
+                    }
+                    Err(error) => {
+                        const ERR_MSG: &'static str =
+                            "failed to prepare instance-start saga";
+                        error!(
+                            opctx.log,
+                            "{ERR_MSG} for {instance_id}";
+                            "instance_id" => %instance_id,
+                            "error" => %error,
+                        );
+                        status
+                            .restart_errors
+                            .push((instance_id, format!("{ERR_MSG}: {error}")))
+                    }
+                };
+            }
+
+            let total_sagas_started = tasks.len();
+            debug!(
+                opctx.log,
+                "found instance in need of reincarnation";
+                "instances_found" => found,
+                "total_found" => status.instances_found,
+                "sagas_started" => total_sagas_started - prev_sagas_started,
+                "total_sagas_started" => total_sagas_started,
+            );
+        }
+
+        // All sagas started, wait for them to come back...
+        while let Some(saga_result) = tasks.join_next().await {
+            match saga_result {
+                // Start saga completed successfully
+                Ok(Ok(instance_id)) => {
+                    debug!(
+                        opctx.log,
+                        "welcome back to the realm of the living, {instance_id}!";
+                        "instance_id" => %instance_id,
+                    );
+                    status.instances_reincarnated.push(instance_id);
+                }
+                // The instance was restarted by another saga, that's fine...
+                Ok(Err((instance_id, Error::Conflict { message })))
+                    if message.external_message()
+                        == instance_start::ALREADY_STARTING_ERROR =>
```

Comment:
We usually avoid relying on specific error messages.  Is it safe to assume that if you got `Error:Conflict`, you're in this case?

Reply:
I, also, felt some amount of distaste for inspecting the error message. But, I
wasn't totally confident that _any_ `Error::Conflict` just meant that the
instance had changed state --- I thought it was possible that one could be
propagated up from one of the sled-agent API calls or something, and figured that might indicate a
more interesting failure than the one from the DB query that moves the instance
to `Starting`. It might be fine though.

Reply:
Hmm, digging through the `instance-start` saga, I _think_ you're right and we
can handle any `Error::Conflict` here.